src/xdp-utils.c

```diff
@@ -454,133 +453,91 @@ parse_app_info_from_flatpak_info (int pid, GError **error)
   return g_steal_pointer (&app_info);
 }
 
-static gboolean
-aa_is_enabled (void)
-{
-  static int apparmor_enabled = -1;
-  struct stat statbuf;
-  struct mntent *mntpt;
-  FILE *mntfile;
-
-  if (apparmor_enabled >= 0)
-    return apparmor_enabled;
-
-  apparmor_enabled = FALSE;
-
-  mntfile = setmntent ("/proc/mounts", "r");
-  if (!mntfile)
-    return FALSE;
-
-  while ((mntpt = getmntent (mntfile)))
-    {
-      g_autofree char *proposed = NULL;
-
-      if (strcmp (mntpt->mnt_type, "securityfs") != 0)
-        continue;
-
-      proposed = g_strdup_printf ("%s/apparmor", mntpt->mnt_dir);
-      if (stat (proposed, &statbuf) == 0)
-        {
-          apparmor_enabled = TRUE;
-          break;
-        }
-    }
-
-  endmntent (mntfile);
-
-  return apparmor_enabled;
-}
-
-#define UNCONFINED		"unconfined"
-#define UNCONFINED_SIZE		strlen(UNCONFINED)
-
-static gboolean
-parse_unconfined (char *con, int size)
-{
-  return size == UNCONFINED_SIZE && strncmp (con, UNCONFINED, UNCONFINED_SIZE) == 0;
-}
-
 static char *
-aa_splitcon (char *con, char **mode)
+freezer_cgroup_for_pid (int pid, GError **error)
 {
-  char *label = NULL;
-  char *mode_str = NULL;
-  char *newline = NULL;
-  int size = strlen (con);
-
-  if (size == 0)
-    return NULL;
+  g_autofree char *cgroup_path = NULL;
+  g_autofree char *contents = NULL;
+  g_autoptr(GError) local_error = NULL;
+  g_auto(GStrv) lines = NULL;
+  int i;
 
-  /* Strip newline */
-  if (con[size - 1] == '\n')
+  cgroup_path = g_strdup_printf ("/proc/%u/cgroup", pid);
+  if (!g_file_get_contents (cgroup_path, &contents, NULL, &local_error))
     {
-      newline = &con[size - 1];
-      size--;
+      /* Don't propagate error if system does not support cgroups */
+      if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        g_propagate_error (error, g_steal_pointer (&local_error));
```

Comment:
Its a bit weird to return NULL and not set error, maybe should comment this at the top of the function.

Reply:
I was using the pattern of `parse_app_info_from_flatpak_info`, to separate out the case of an expected failure when the kernel feature is not available from the cases we expect to succeed but encounter an error.  I've added a comment to make this clear.